### PR TITLE
Add focus trap to the RHP (v2)

### PR DESCRIPTION
**Pull request by kosmydel**
### Details

This is a new version of [this PR](https://github.com/Expensify/App/pull/24316). The previous one was reverted due to two bugs:
- https://github.com/Expensify/App/issues/27519
- https://github.com/Expensify/App/issues/27516

This PR addresses the above issues and reintroduces the `FocusTrap`. I've updated the test plan and posted new videos to cover the above issues.

### Fixed Issues

$ https://github.com/Expensify/App/issues/15631
PROPOSAL: https://github.com/Expensify/App/issues/15631#issuecomment-1641623988

### Tests

- [ ] All the _QA Steps_ listed below.
- [ ] Remove children elements from the RHP panel (in the `InitialSettingsPage`), verify that the focus trap works without focusable elements inside

### Offline tests

Doesn't change anything network-related.

### QA Steps

- [ ] Open RHP modals:
  - [ ] Open Settings (press user avatar in the top left corner, above the report's list),
  - [ ] In the report press user avatar to open user details
  - [ ] In the report press the header (name of the report) to open report details
  - [ ] In Settings go to Workspaces, and then press "New workspace" button.
  - [ ] Try to cover as many RHP panels as possible.
- [ ] Verify that no errors appear in the JS console
- [ ] On each modal listed above verify (web + desktop):
  - [ ] Press **Space** - verify that you can't scroll in the background
  - [ ] Navigate using **Tab** and **Shift+Tab**, verify that you can't focus outside the modals.
  - [ ] Verify that you can select the focused element using Enter.
  - [ ] Do the above points for subpages (i.g. enter Settings -> Profile).
  - [ ] Refresh the page with the RHP opened, and verify that everything works.
  - [ ] Click outside the modal to check if it closes.
  - [ ] Click Escape and check if the modal closes (and everything works).
- [ ] Other tests (web + desktop):
  - [ ] With the RHP panel hidden verify that you can change focus using the (Shift+)Tab correctly.
  - [ ] On the Settings RHP resize windows so it is small enough so the scroll bar appears. Now focus on some button in the scroll view, and then try to scroll using the (Shift+)Space.
  - [ ] Verify focus trap behavior for accessibility.
- [ ] On each modal (all platforms):
  - [ ] Verify that you can scroll the modals
  - [ ] Navigate in modals, go to subpages, and verify that everything works
- [ ] Open the search panel:
  - [ ] Verify that the autofocus works, and the input bar is automatically focused.
- [ ] Go around the app, and check if everything works.
- [x] [Test 2FA page crashing issue](https://github.com/Expensify/App/issues/27519):
  - [ ] Log in to an account with the 2FA configured
  - [ ] Go to Settings -> Security -> Two-factor authentication
  - [ ] Verify that the App doesn't crash
  - [ ] Test full 2FA flow:
    - [ ] add 2FA to the account,
    - [ ] remove 2FA from the account.
  - [ ] Especially verify the pages with `AnimatedStepProvider`, currently it is only the `TwoFactorAuthPage`.
- [ ] [Test search field remain focused after CTRL + K and CTRL + J issue](https://github.com/Expensify/App/issues/27516)
  - [ ] Press CTRL + K, then CTRL + J - check if focus trap works as expected:
    - [ ] The search field in the background shouldn't remain focused,
    - [ ] After pressing escape with the modal open the search field should focus again
    - [ ] After clicking outside the modal the search field should focus again
  - [ ] Do the same tests from the above in CTRL + SHIFT + K, CTRL + J
  - [ ] Check Settings:
    - [ ] Open Settings
    - [ ] With tab focus on some button (e.g.. Profile)
    - [ ] Press CTRL + J to open the modal
    - [ ] ~~The background button should blur~~ (the modal was changed)
    - [ ] Exit the modal
    - [ ] ~~Verify that the button is focused again~~
  - [ ] Verify other modals, check if work:
    - [ ] Add attachment to the report,
    - [ ] Go to Settings -> Profile -> Personal details -> Home Address -> Country, press CTRL + J, check if everything is as expected
    - [ ] Go to Settings -> Profile -> Upload new avatar (modal) -> Open CTRL + J, check if works
    - [ ] Press the "New" button (green cross at the bottom), then CTRL + J
  - [ ] Create a new account. Go to Settings. Verify everything works.
  - [ ] Go to the Settings -> Security -> Close account -> enter default contact -> Click `Close acctount`. Verify if the focus trap is working in the confirmation modal

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section.
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

**Fixed:**

https://github.com/Expensify/App/assets/104823336/13d462ee-5273-44fe-b870-dbee1a91d1f3

https://github.com/Expensify/App/assets/104823336/f458788b-7bd6-4d4e-8ce5-98e9af50a222

Workspace creation:

https://github.com/Expensify/App/assets/104823336/12fd737d-ea34-4c7c-aed3-63e1e27b3712

**Old version:**
https://github.com/Expensify/App/assets/104823336/4521b3e2-b25c-427c-a1e6-5294242e39e1

</details>

<details>
<summary>Mobile Web - Chrome</summary>

**Fixed:**

https://github.com/Expensify/App/assets/104823336/b6806e86-b2d8-42a4-9b6b-0fe863bf02f2

https://github.com/Expensify/App/assets/104823336/505add35-a5e2-4408-a614-15394b9bdcc8

</details>

<details>
<summary>Mobile Web - Safari</summary>

**Fixed:**

https://github.com/Expensify/App/assets/104823336/b6372896-c94c-4a3f-9c23-5b148bd775e5

https://github.com/Expensify/App/assets/104823336/be80d170-144e-4054-a1da-5fb098fdc78d

</details>

<details>
<summary>Desktop</summary>

**Fixed:**

https://github.com/Expensify/App/assets/104823336/6c138fdd-9f65-4a21-83cc-e2162ee59e8f

https://github.com/Expensify/App/assets/104823336/83ceca8e-e8f1-4243-821c-13325e75e4bc

Workspace creation:

https://github.com/Expensify/App/assets/104823336/d208d055-755a-4c3c-8b7f-3aefbbb8f6ce

**Old version:**
https://github.com/Expensify/App/assets/104823336/43e3570d-5293-4804-890d-c8675f25d934

</details>

<details>
<summary>iOS</summary>

**Fixed:**

https://github.com/Expensify/App/assets/104823336/f6145803-b8ad-4ac8-a44f-c29849f409dc

https://github.com/Expensify/App/assets/104823336/09dd8142-9760-45ba-b230-a70f7013ff73

</details>

<details>
<summary>Android</summary>

**Fixed:**

https://github.com/Expensify/App/assets/104823336/280e05f7-364d-44fe-bdf6-7b6125a8fc36

https://github.com/Expensify/App/assets/104823336/c896ce9e-e4d6-474a-a776-b758cee35cd8

</details>

### CODE REVIEW

**Comment by mountiny** [comment]
I want to flag with this PR https://github.com/Expensify/App/pull/27230#issuecomment-1722245821 we faced some issues with the focusTrap. Lets make sure we wait for that and we address it here.

**Comment by kosmydel** [comment]
Hey @mountiny, can we remove HOLD from this as the other PR was merged?

I merged the main and resolved conflicts but didn't thoroughly test it yet (as it is on hold).

**Comment by mountiny** [comment]
@kosmydel thank you! Please go ahead and test away

**Comment by kosmydel** [comment]
Hey @fedirjh, I tested it locally (including testing two regressions fixed) and it looks like everything works. I think we can proceed with the review.

cc @mountiny

**Comment by mountiny** [comment]
All yours @fedirjh

**Comment by kosmydel** [comment]
I've resolved merged & resolved conflicts. I have to test it thoroughly again but will do it tomorrow. Marking as draft

**Comment by kosmydel** [comment]
Hey, @fedirjh I've tested it, and it looks like it is ready for review

**Comment by kosmydel** [comment]
> @kosmydel Could you please merge main and resolve the conflicts?

Done @fedirjh :)

**Review by fedirjh — APPROVED** [review]
Looks good to me.

File: src/pages/settings/Security/TwoFactorAuth/TwoFactorAuthSteps.js [hunk 1]
```diff
@@ -13,24 +13,22 @@ import TwoFactorAuthContext from './TwoFactorAuthContext';
 import {defaultAccount, TwoFactorAuthPropTypes} from './TwoFactorAuthPropTypes';
 
 function TwoFactorAuthSteps({account = defaultAccount}) {
-    const [currentStep, setCurrentStep] = useState(CONST.TWO_FACTOR_AUTH_STEPS.CODES);
+    const calculateCurrentStep = useMemo(() => {
```
Comment (roryabraham) [inline src/pages/settings/Security/TwoFactorAuth/TwoFactorAuthSteps.js:16]: There is some weirdness going on in this component, essentially derived state. At minimum, I'd rename `calculateCurrentStep` to `currentStepDerivedFromProps` and move the related `useEffect` up:

```diff
diff --git a/scripts/start-android.sh b/scripts/start-android.sh
old mode 100644
new mode 100755
diff --git a/src/pages/settings/Security/TwoFactorAuth/TwoFactorAuthSteps.js b/src/pages/settings/Security/TwoFactorAuth/TwoFactorAuthSteps.js
index 116d44e9e3..0dfc6ef951 100644
--- a/src/pages/settings/Security/TwoFactorAuth/TwoFactorAuthSteps.js
+++ b/src/pages/settings/Security/TwoFactorAuth/TwoFactorAuthSteps.js
@@ -13,23 +13,21 @@ import TwoFactorAuthContext from './TwoFactorAuthContext';
 import {defaultAccount, TwoFactorAuthPropTypes} from './TwoFactorAuthPropTypes';
 
 function TwoFactorAuthSteps({account = defaultAccount}) {
-    const calculateCurrentStep = useMemo(() => {
+    const currentStepDerivedFromProps = useMemo(() => {
         if (account.twoFactorAuthStep) {
             return account.twoFactorAuthStep;
         }
         return account.requiresTwoFactorAuth ? CONST.TWO_FACTOR_AUTH_STEPS.ENABLED : CONST.TWO_FACTOR_AUTH_STEPS.CODES;
     }, [account.requiresTwoFactorAuth, account.twoFactorAuthStep]);
-
-    const [currentStep, setCurrentStep] = useState(calculateCurrentStep);
+    const [currentStep, setCurrentStep] = useState(currentStepDerivedFromProps);
+    useEffect(() => {
+        setCurrentStep(currentStepDerivedFromProps);
+    }, [currentStepDerivedFromProps]);
 
     const {setAnimationDirection} = useAnimatedStepContext();
 
     useEffect(() => () => TwoFactorAuthActions.clearTwoFactorAuthData(), []);
 
-    useEffect(() => {
-        setCurrentStep(calculateCurrentStep);
-    }, [calculateCurrentStep]);
-
     const handleSetStep = useCallback(
         (step, animationDirection = CONST.ANIMATION_DIRECTION.IN) => {
             setAnimationDirection(animationDirection);
```

However, I think a simpler approach would be to just use props and rely on Onyx to perform the update:

```diff
diff --git a/scripts/start-android.sh b/scripts/start-android.sh
old mode 100644
new mode 100755
diff --git a/src/pages/settings/Security/TwoFactorAuth/TwoFactorAuthSteps.js b/src/pages/settings/Security/TwoFactorAuth/TwoFactorAuthSteps.js
index 116d44e9e3..4d342c8dac 100644
--- a/src/pages/settings/Security/TwoFactorAuth/TwoFactorAuthSteps.js
+++ b/src/pages/settings/Security/TwoFactorAuth/TwoFactorAuthSteps.js
@@ -13,28 +13,22 @@ import TwoFactorAuthContext from './TwoFactorAuthContext';
 import {defaultAccount, TwoFactorAuthPropTypes} from './TwoFactorAuthPropTypes';
 
 function TwoFactorAuthSteps({account = defaultAccount}) {
-    const calculateCurrentStep = useMemo(() => {
+    const currentStep = useMemo(() => {
         if (account.twoFactorAuthStep) {
             return account.twoFactorAuthStep;
         }
         return account.requiresTwoFactorAuth ? CONST.TWO_FACTOR_AUTH_STEPS.ENABLED : CONST.TWO_FACTOR_AUTH_STEPS.CODES;
     }, [account.requiresTwoFactorAuth, account.twoFactorAuthStep]);
 
-    const [currentStep, setCurrentStep] = useState(calculateCurrentStep);
-
     const {setAnimationDirection} = useAnimatedStepContext();
 
     useEffect(() => () => TwoFactorAuthActions.clearTwoFactorAuthData(), []);
 
-    useEffect(() => {
-        setCurrentStep(calculateCurrentStep);
-    }, [calculateCurrentStep]);
-
     const handleSetStep = useCallback(
         (step, animationDirection = CONST.ANIMATION_DIRECTION.IN) => {
             setAnimationDirection(animationDirection);
             TwoFactorAuthActions.setTwoFactorAuthStep(step);
-            setCurrentStep(step);
         },
         [setAnimationDirection],
     );
```

Note that `TwoFactorAuthActions.setTwoFactorAuthStep` will immediately call `Onyx.merge` with the new step, then thanks to `withOnyx` the component will re-render with the new `twoFactorAuthStep` prop, and `currentStep` will be recomputed accordingly.

**Review by roryabraham — CHANGES_REQUESTED** [review]

File: src/components/FocusTrapView/index.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+     * Focus trap always needs a focusable element.
+     * In case that we don't have any focusable elements in the modal,
+     * the FocusTrap will use fallback View element using this ref.
+     */
+    const ref = useRef(null);
+
+    /**
+     * We have to set the 'tabindex' attribute to 0 to make the View focusable.
+     * Currently, it is not possible to set this through props.
+     * After the upgrade of 'react-native-web' to version 0.19 we can use 'tabIndex={0}' prop instead.
```
[the remark this answers is not in the document]
Comment (kosmydel) [reply src/components/FocusTrapView/index.js:46]: Sure, we can wait for this.

[the remark this answers is not in the document]
Comment (kosmydel) [reply src/components/FocusTrapView/index.js]: I've updated the PR with this change.

**Comment by kosmydel** [comment]
I've updated to use `react-native-web` `tabIndex` property, but can't test it thoroughly, since there was a regression: the user can't navigate between pressables using the tab. 
[More info here.](https://github.com/Expensify/App/issues/31098#issuecomment-1803633131) I think we should wait until they resolve it.

cc @roryabraham @fedirjh

**Comment by kosmydel** [comment]
recheck

**Comment by kosmydel** [comment]
We have to wait for https://github.com/Expensify/App/pull/31233 to be merged to resolve the PR Author Checklist check failing.

**Comment by roryabraham** [comment]
Sorry for the inconvenience – https://github.com/Expensify/App/pull/31233 has been merged

**Comment by kosmydel** [comment]
> @kosmydel We still have a failing test 😕

@fedirjh Yeah, I see it, but I'm not sure what can I do about it :( I do not have access to the `snyk` account. 

cc @roryabraham, can you have a look what is the reason behind the `snyk` test failing?

---

I've also migrated the `FocusTrapView` to the TypeScript because the `Modal` component was migrated, and I couldn't proceed without that. cc @blazejkustra

**Comment by blazejkustra** [comment]
I would like to ask at least one of you @fabioh8010 / @kubabutkiewicz to take a look at this PR; especially `src/components/FocusTrapView/index.tsx`. The ref is of type `HTMLElement`, when it's passed to the View component we need to assert because of that:
```ts
            <View
                ref={ref as unknown as React.RefObject<View>}
```

Do you have other ideas how to improve it?  How about augmenting `View`'s `ref` type in `src/types/modules/react-native`?

**Comment by fabioh8010** [comment]
Hi @blazejkustra , after trying some approaches here is my investigation:

First, I tried using module augmentation as suggested but I think we won't be able to change `ref` type properly. It looks like `ref` type is supplied by generics and is tied to the component itself (in this case, `View`).

Screenshot 2023-11-16 at 16 10 56

I overrode `ref` in ViewProps:

```ts
    interface ViewProps extends WebViewProps {
        // also tried ref?: React.LegacyRef<View> | React.LegacyRef<HTMLDivElement>;
        ref?: React.LegacyRef<View | HTMLDivElement>;
    }
```

But by doing so we end up having TS issues in the `ref` prop.

Screenshot 2023-11-16 at 16 20 39

It's worth to mention that module augmentation doesn't work well when we need to override the type of properties that already exists, sometimes the property type is not changed at all.

---

One solution that we could follow is **isolate the type issue**, somewhat I've done with some style helpers that suffer from the same module augmentation problems, like [overflowAuto](https://github.com/Expensify/App/blob/main/src/styles/utilities/overflowAuto/index.ts).

We can create a helper that accepts both `View` and `HTMLDivElement` refs, and return the same object typed as `View` ref.

```ts
const viewRef = (ref: React.RefObject<View | HTMLDivElement>) => ref as React.RefObject<View>;
```

Also, let's add `null` to `useRef` to get a `RefObject` instead of `MutableRefObject`.

```ts
const ref = useRef<HTMLDivElement>(null);
```

And finally, we use `viewRef` in the `View` component, thus ensuring it will satisfy the type requirement.

```ts
            <View
                ref={viewRef(ref)}
                tabIndex={0}
                // eslint-disable-next-line react/jsx-props-no-spreading
                {...props}
            />
```

Entire solution:

```ts
const viewRef = (ref: React.RefObject<View | HTMLDivElement>) => ref as React.RefObject<View>;

function FocusTrapView({isEnabled = true, isActive = true, shouldEnableAutoFocus = false, ...props}: FocusTrapViewProps) {
    /**
     * Focus trap always needs a focusable element.
     * In case that we don't have any focusable elements in the modal,
     * the FocusTrap will use fallback View element using this ref.
     */
    const ref = useRef<HTMLDivElement>(null);

    return isEnabled ? (
        <FocusTrap
            active={isActive}
            focusTrapOptions={{
                initialFocus: () => (shouldEnableAutoFocus && ref.current) ?? false,
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                fallbackFocus: () => ref.current!,
                clickOutsideDeactivates: true,
            }}
        >
            <View
                ref={viewRef(ref)}
                tabIndex={0}
                // eslint-disable-next-line react/jsx-props-no-spreading
                {...props}
            />
        </FocusTrap>
    ) : (
        props.children
    );
}
```

Since `viewRef` returns the exact same object reference, it should be safe to use it in the prop. We can create `viewRef` in a separate utility file to be used in case we face similar situations.

**Comment by kubabutkiewicz** [comment]
I like that idea from @fabioh8010 I had similar problem in this [this](https://github.com/Expensify/App/pull/31012) and did some awkward typing. But I think we should follow @fabioh8010 approach.

**Comment by kosmydel** [comment]
I'm not sure what can I do about the failing Snyk tests. Can you @roryabraham provide me with more details, as I cannot access it?

**Review by fedirjh — APPROVED** [review]
Re-tested and everything is working fine. Not sure what to do about the failing tests.

File: src/components/FocusTrapView/types.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import {ReactNode} from 'react';
+import {ViewProps} from 'react-native';
+
+type FocusTrapViewProps = {
+    /** Children to wrap with FocusTrap */
+    children: ReactNode;
```
Comment (blazejkustra) [inline src/components/FocusTrapView/types.ts:6]: Use `ChildrenProps` from `src/types/utils/ChildrenProps.ts` instead:
```ts
type FocusTrapViewProps = ChildrenProps & { ...
```

File: src/components/FocusTrapView/index.tsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
+/*
+ * The FocusTrap is only used on web and desktop
+ */
+import FocusTrap from 'focus-trap-react';
+import React, {useRef} from 'react';
+import {View} from 'react-native';
+import FocusTrapViewProps from './types';
+
+const viewRef = (ref: React.RefObject<View | HTMLDivElement>) => ref as React.RefObject<View>;
```
Comment (blazejkustra) [inline src/components/FocusTrapView/index.tsx:9]: Move this util to `src/types/utils`

**Review by blazejkustra — CHANGES_REQUESTED** [review]

**Comment by kosmydel** [comment]
I've found two more screens to fix and fixed those [here](https://github.com/Expensify/App/pull/27670/commits/25c51254906543bdb73bcef2c57922a627f63616). cc @fedirjh

**Review by blazejkustra — APPROVED** [review]

**Comment by roryabraham** [comment]
> I'm not sure what can I do about the failing Snyk tests

We have an open support ticket with snyk as these tests are failing on all PRs (at least, all PRs from forks)

**Review by roryabraham — APPROVED** [review]

**Comment by situchan** [comment]
This PR caused above deploy blockers with the same root cause.

**Comment by luacmartins** [comment]
@kosmydel @roryabraham @blazejkustra we're reverting this PR since it caused the following blockers:

- https://github.com/Expensify/App/issues/31687
- https://github.com/Expensify/App/issues/31725
- https://github.com/Expensify/App/issues/31734

Could you please work on the PR and a fix for the issues above?

File: package.json [hunk 5]
```diff
@@ -100,6 +100,7 @@
     "domhandler": "^4.3.0",
     "expensify-common": "git+ssh://git@github.com/Expensify/expensify-common.git#886f90cbd5e83218fdfd7784d8356c308ef05791",
     "fbjs": "^3.0.2",
+    "focus-trap-react": "^10.2.1",
```
Comment (shubham1206agra) [inline package.json:103]: Please use v10.2.3 to include https://github.com/focus-trap/focus-trap/issues/1068

**Comment by kosmydel** [comment]
> @kosmydel @roryabraham @blazejkustra we're reverting this PR since it caused the following blockers:
> 
> * [[$500] IOU - Blue box surrounding the currency after the currency selection #31687](https://github.com/Expensify/App/issues/31687)
> * [ Workspace - Blue focus is shown when changing workspace name #31725](https://github.com/Expensify/App/issues/31725)
> * [BA - Error message doesn't stop displaying instantly after valid data is entered in date field #31734](https://github.com/Expensify/App/issues/31734)
> 
> Could you please work on the PR and a fix for the issues above?

Hey, sure I will check it. However, currently, I am focusing on the ideal nav project, as it has the highest priority.

**Comment by kosmydel** [comment]
> @kosmydel any new updates ?

Hey, I hope that next week I will have some more time to look at this.

**Pull request merged**